### PR TITLE
[WIP] Dynamic pool configuration

### PR DESCRIPTION
**Pull request by michalmuskala**
This is a proof of concept of how we could allow for more dynamic configuration of repo.

Here's what's happening:
* A new `Ecto.Registry` module for a simple process-attached key-value store.
* A new adapter callback `init/2`, where adapter returns "runtime data", that is later stored in the registry
* Instead of passing the repo module, all adapter callbacks receive this "runtime data"
* The `__pool__` definition is removed
* The `repo` passed to various functions in the `Ecto.Repo.*` modules is treated as a process or a process name. They use it to retrieve data from the registry and pass to the adapter callbacks in place of repo.

This still needs work to truly allow dynamic repos, but should server as a good starting point for further work. Only tested with postgres for now.

Closes https://github.com/elixir-ecto/ecto/issues/1964

\cc @josevalim @talentdeficit

### CODE REVIEW

File: lib/ecto/adapter.ex [hunk 1]
```diff
@@ -45,6 +45,11 @@ defmodule Ecto.Adapter do
   """
   @callback child_spec(repo, options) :: Supervisor.Spec.spec
 
+  @doc """
+  Initializes the adapter
+  """
+  @callback init(repo, options) :: {:ok, state :: term} | {:error, reason :: term}
```
Comment (josevalim) [inline lib/ecto/adapter.ex]: There is no need for an init callback. If the adapter needs to do extra work, then it needs to wrap the child_spec or the function in the child_spec returned by DBConnection by its function/child_spec and then execute any custom action that is necessary.

File: lib/ecto/repo/queryable.ex [hunk 2]
```diff
@@ -11,11 +11,13 @@ defmodule Ecto.Repo.Queryable do
   require Ecto.Query
 
   def transaction(adapter, repo, fun, opts) when is_function(fun, 0) do
-    adapter.transaction(repo, opts, fun)
+    data = Ecto.Registry.lookup(repo)
```
Comment (josevalim) [inline lib/ecto/repo/queryable.ex]: I am not sure this is the way we want to go. This is imposing all adapters to go through the Ecto.Registry, which means that now all adapters need to register something when they are initialized. I would prefer if, as this change relates only to pools, we did not change the adapter API at all, but only the SQL adapters which are the ones who effectively use a pool.

**Comment by josevalim** [comment]
@michalmuskala I believe this PR is trying to do too much at once. I would prefer if we originally tackled the `__pool__` issue separately from the dynamic repositories. That's because pools are not a requirement for every Ecto adapter and if we put those together we will implicitly require all adapters to use a pool (which this PR seems to be accidentally doing).

In as few steps as possible, here is how I would see this PR working:

  1. It uses a registry
  2. It overrides the child_spec to invoke a function that would allow the pool options to be registered on start up
  3. We change __pool__ to lookup the pool information in the registry instead of being compiled in memory

This way we keep the existing APIs and make almost everything dynamic. The only non-dynamic piece is the query cache table which we can then discuss how to tackle next.

I would also suggest for us to use Elixir's v1.4 registry. Ecto 2.2 is not coming out soon and by the time it is out, I assume requiring Elixir v1.4 won't be a huge issue.

I would personally prefer if we don't do the lookup and let adapters be the ones responsible for doing any registry lookup. All we guarantee is that the adapter will either receive the repo name or a pid. We can then provide an up and running registry as a convenience (so an adapter doesn't need to start one).

**Comment by michalmuskala** [comment]
This is exactly the way I tried to do it initially, but it presented a lot of issues. Most notably, we need to associate the top-level repo supervisor to the pool config, and not the pool process itself - if we had the pid of the pool process, we wouldn't need any of this.

**Comment by josevalim** [comment]
Good call.

You should be able to get the pid though if you do the child_spec trick I
mentioned, because you will be the one calling DBConnection's start_link.

However this means you won't be able to use Elixir registry. So your
registry implemention will still be necessary.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by michalmuskala** [comment]
@josevalim updated. The pool handling is now entirely on the adapter side - this actually simplified the code.

The only thing, I'm not 100% convinced about is that we assume in the `child_spec` function of the adapter that it will be called from within the main repo supervisor. I guess this is fine, but wanted to point it out. Ideally the supervisor would pass the pid/name to the `child_spec` function, but this would require changing the adapter API.

The ets table shouldn't be a problem - we can always name it as the top level repo supervisor, and require it to be named with an atom (could be called differently that the repo module itself - by default it wouldn't be called differently).

I was wondering how to provide the dynamic repo interface - I see two ways:
* add a module `Ecto.Repo.Dynamic` that exports all the functions `Ecto.Repo` does, but adds additional two arguments - for the name/pid of the dynamic repo.
* make all functions in the regular repos, accept an additional argument/option for the repo process.

I lean towards the first solution.

File: lib/ecto/adapters/sql.ex [hunk 3]
```diff
@@ -525,26 +523,26 @@ defmodule Ecto.Adapters.SQL do
 
   @doc false
   def reduce(repo, statement, params, mapper, opts, acc, fun) do
-    {pool, default_opts} = repo.__pool__
+    {repo_mod, pool, default_opts} = Ecto.Registry.lookup(repo)
```
Comment (josevalim) [inline lib/ecto/adapters/sql.ex]: We should probably move `Ecto.Registry.lookup(repo)` to a private function (given you had to repeat it in many places in this PR).

File: lib/ecto/repo/supervisor.ex [hunk 4]
```diff
@@ -111,8 +111,8 @@ defmodule Ecto.Repo.Supervisor do
           :ets.new(repo, [:set, :public, :named_table, read_concurrency: true])
         end
         supervise(children, strategy: :one_for_one)
-      :ignore ->
```
Comment (josevalim) [inline lib/ecto/repo/supervisor.ex]: Why this change?

File: test/ecto/adapters/sql_test.exs [hunk 5]
```diff
@@ -1,30 +1,30 @@
-defmodule Ecto.Adapters.SQLTest do
```
Comment (josevalim) [inline test/ecto/adapters/sql_test.exs]: We can delete this file but we need to add an integration test that shows we can pass the pool options on `start_link`.

Comment (josevalim) [inline test/ecto/adapters/sql_test.exs]: One good option to test is the `:pool_name` option.

**Comment by josevalim** [comment]
> The only thing, I'm not 100% convinced about is that we assume in the child_spec function of the adapter will be called from within the main repo supervisor.

That's a reasonable assumption to make. In our proposal in the elixir-lang-core mailing list we are pushing towards the same direction.

> add a module Ecto.Repo.Dynamic that exports all the functions Ecto.Repo does, but adds additional argument - for the name/pid of the dynamic repo.

This is my preference too. We can also consider adding those functions directly to `Ecto.Repo` but it may end-up cluttering the API. If not `Ecto.Repo`, I would also prefer to call it `Ecto.DynamicRepo`.

**Comment by josevalim** [comment]
@michalmuskala I pushed a commit that slightly renames normalize_config to pool_config and cleans up some of the indirection.

**Comment by josevalim** [comment]
I went ahead and fixed the last issues beacuse I want to do more work on top of this. Thanks @michalmuskala!

**Comment by michalmuskala** [comment]
:+1: sorry for the delay

**Pull request merged**